### DISCUSSION TITLE
Tools to migrate from Mnesia

### DISCUSSION THREAD

**Discussion by dumbbell** [rabbitmq/khepri#87] [Ideas]
I think it would be useful to provide a module to help users migrate from Mnesia to Khepri. Perhaps the opposite too to make things even easier.

Here are some ideas:

*   We could provide a module to fill a Khepri store with the data from Mnesia tables. The layout could be something like that:
    ```erlang
    RecordIndexPositions = mnesia:table_info(MnesiaTableName, index),
    
    %% For each record in table `MnesiaTableName', we need to decide on the record
    %% ID to put in the path. It could be limited to Mnesia tables with a single
    %% index. Also the type could be limited to atoms and binaries, or converted to
    %% a binary.
    RecordId = compute_record_id(Record, RecordIndexPositions)
    Path = [MnesiaTableName, RecordId],
    khepri:put(StoreId, Path, Record).
    ```

    We already have some interesting things in the RabbitMQ integration of Khepri. Perhaps we can move some of this code to Khepri itself.

*   We could let the user decide how to derive a path from a record, either by taking an anonymous function or by leveraging ETS matching specs which allow to format the output of a match:
    ```erlang
    PathMatchSpec = {
      #record{name = '$1', email = '$2', _ = '_'},
      [],
      [from_mnesia, MnesiaTableName, '$1', '$2']}.
    ```

*   We could provide a module with an API close to Mnesia API. Probably difficult, but something as close  as possible to a drop-in replacement to `mnesia` calls.

**Comment by michaelklishin** [comment]
It would be interesting to consider a module that transfers data from a Mnesia table into a Khepri "directory" (set of keys). But I see that this is to reduce code changes needed (our team obviously has a lot of vested interest here), which is more important.
